server/src/main/java/io/crate/replication/logical/LogicalReplicationService.java

```diff
@@ -301,44 +344,125 @@ public void verifyTablesDoNotExist(String subscriptionName,
+                } else {
+                    updateSubscriptionState(
+                        subscriptionName,
+                        Subscription.State.FAILED,
+                        "Failed restoring subscription, error=" + err.getMessage()
+                    ).whenComplete((ignored, ignoredErr) -> finalFuture.completeExceptionally(err));
+                }
             }
+        );
 
-            @Override
-            protected void doRun() {
-                restoreService.restoreSnapshot(restoreRequest, future);
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(
+            () -> {
+                try {
+                    restoreService.restoreSnapshot(restoreRequest, restoreFuture);
+                } catch (Exception e) {
+                    restoreFuture.onFailure(e);
+                }
             }
-        });
-        return future;
+        );
+
+        return finalFuture;
     }
 
-    private CompletableFuture<Boolean> afterReplicationStarted(RestoreService.RestoreCompletionResponse response) {
-        var future = new FutureActionListener<RestoreSnapshotResponse, Boolean>(
-            restoreSnapshotResponse -> {
-                RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
-                if (restoreInfo == null) {
-                    LOGGER.error(
-                        "Restore failed, restoreInfo = NULL, seems like a master failure happened while restoring");
-                    return false;
-                } else if (restoreInfo.failedShards() == 0) {
-                    LOGGER.debug("Restore success, following will start once shards are active");
-                    return true;
+    private CompletableFuture<Boolean> afterReplicationStarted(String subscriptionName,
+                                                               RestoreService.RestoreCompletionResponse response) {
+        var future = new CompletableFuture<Boolean>();
+        var restoreFuture = new FutureActionListener<RestoreSnapshotResponse, RestoreSnapshotResponse>(
+            Function.identity());
+        restoreFuture.whenComplete(
+            (restoreSnapshotResponse, err) -> {
+                if (err == null) {
+                    RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
+                    if (restoreInfo == null) {
+                        LOGGER.error(
+                            "Restore failed, restoreInfo = NULL, seems like a master failure happened while restoring");
+                        updateSubscriptionState(
+                            subscriptionName,
+                            Subscription.State.FAILED,
+                            "Error while initial restoring the subscription relations"
+                        ).whenComplete(
+                            (ignore, ignoredErr) -> future.complete(false)
+                        );
+                    } else if (restoreInfo.failedShards() == 0) {
```

Comment:
Similar here - might be possible to use the return values of `whenComplete` to do the chaining, to reduce the number of indentations a bit?

Reply:
Same here, results of the `updateSubscriptionState` are ignored. I agree that this is not great to read, but I could find a better way.